src/utils/styled_lines_iterator.rs

```diff
@@ -0,0 +1,224 @@
+use utils::LinesIterator;
+use style::*;
+
+use unicode_width::UnicodeWidthStr;
+
+/// Generates rows of styled text in constrained width.
+///
+/// Given styled text and a width constraint, it iterates over
+/// collections of styled spans, each making up a row within the constraint.
+///
+/// # Examples
+///
+/// ```
+/// # use cursive::style::{StyledString, StyledText};
+/// # use cursive::utils::StyledLinesIterator;
+/// # let text_styled: &[StyledString] = &[].as_ref();
+/// let text_plain = text_styled.to_plain();
+/// let spans = StyledLinesIterator::make_spans(text_styled);
+/// // You can now keep `text_plain` and `spans` alive while using `iter`.
```

Comment:
Excuse me, I have a little trouble understanding your idea.

> Since we can now print content which is not directly present in the input (because of escaping), the indices from StyledSpan are no longer a great solution. Keeping a Cow in the StyledSpan sounds easier.
> Actually, I think we could merge StyledSpan and StyledString. Both refer to a "styled" section from the input text, potentially "rewritten" to account for escaping (hence the Cow).

If the indices are removed, `StyledSpan` contains only the `style` and `width`. Since `StyledString` already contains both of them (`width` via `styled_string.0.as_ref().width()`) merging them would just mean that `StyledSpan` vanishes.

`compute_rows` is basically equivalent to `styled_lines_iter.collect()`, it just solves the ownership problem of `StyledLinesIterator` by removing the ability to iterate, keeping the ownership in the parent scope.

However I also don't like the potential for misuse that `StyledLinesIterator` poses.

Therefore I would suggest to do perform the m stillerge of `StyledSpan` and `StyledString`, in effect just removing `StyledSpan`, and making `StyledLinesIterator` split `StyledString`s in half on line breaks. Since it now uses `Cow` that should be possible. We gain:

- no more potential misuse of `StyledLinesIterator`
- no more numerical indices tricking the borrow checker
- no `text_plain` to be kept alive by the user, it now "lives" in the `Cow`s of a `[StyledString]`

I'll try this so we can see how it compares with your approach. Maybe I missed something, especially since I'm not sure exactly what you mean.